### ISSUE TITLE
Qustions: Softmax for Multi-Class

### ISSUE THREAD

**Issue by gjbit** [qubvel/segmentation_models.pytorch#157]
Hi! Thank you for your work!
I have a question: Is it necessary to have a background class when using softmax activation? So that every pixel belongs to a label.
And what is the difference between 'softmax' and 'softmax2d'?

Thank U!

**Comment by SinclairHudson** [comment]
It is required that you have a background class. Think of softmax as a function that takes a list of numbers and turns them all into probabilities. Check out https://pytorch.org/docs/stable/nn.html?highlight=softmax#torch.nn.Softmax for the actual equation. Importantly, all the probilities **must** sum to 1. If your network thinks a pixel is background, it can't just weight all the other classes zero, because all the values must sum to 1. So you need a background class so that the network can actually predict that class. From what I can see from the PyTorch documentation, Softmax2d is a special case of softmax where the dimension you're applying the softmax over is the second one. This is the most common case in semantic segmentation, where your output is (B x C x W x H). You want to apply the softmax function over the C dimension, the channel dimension, so softmax2d would work for you. Alternatively, softmax(dim=1) would achieve the exact same thing. I don't know why both are included in this repo, because they seem to do the same thing in image segmentation.

**Comment by gjbit** [comment]
> 
> 
> It is required that you have a background class. Think of softmax as a function that takes a list of numbers and turns them all into probabilities. Check out https://pytorch.org/docs/stable/nn.html?highlight=softmax#torch.nn.Softmax for the actual equation. Importantly, all the probilities **must** sum to 1. If your network thinks a pixel is background, it can't just weight all the other classes zero, because all the values must sum to 1. So you need a background class so that the network can actually predict that class. From what I can see from the PyTorch documentation, Softmax2d is a special case of softmax where the dimension you're applying the softmax over is the second one. This is the most common case in semantic segmentation, where your output is (B x C x W x H). You want to apply the softmax function over the C dimension, the channel dimension, so softmax2d would work for you. Alternatively, softmax(dim=1) would achieve the exact same thing. I don't know why both are included in this repo, because they seem to do the same thing in image segmentation.

Thank you very much!!

**Comment by bao18** [comment]
@SinclairHudson, many thanks for your explanation.
I have a very simple follow-up question. In the Keras implementation example, the background is added at the last position (index=-1) of the one-hot encoded of the classes. I wonder if this is always the right way. Can we put the background in the first position (index=0) or at any position?

**Comment by SinclairHudson** [comment]
@bao18 The ordering of the classes in the one-hot encoding is arbitrary; the network will learn which positions are related to which class. However, make sure the ordering is consistent thoughout your whole project, otherwise you may be interpretting your results incorrectly. I personally always keep my "background" class in position 0, because it makes the most sense to me.

**Comment by julienguegan** [comment]
@SinclairHudson Hello, I am kind of struggling to understand for multiclass segmentation if you necessarily need to one-hot encode your mask (CxHxW fill with 0 or 1 value) or not (HxW fill with 0 to C value). Do you have some infos about that ? Thank you !

**Comment by azkalot1** [comment]
As far as I remember pytorch implementation, HxW fill with 0 to C value is the way to go

**Comment by SinclairHudson** [comment]
@julienguegan yes, you would need to one-hot encode your mask for multi-class segmentation. If you had your model predict HxW , where every value was in [0, C], your network wouldn't learn effectively. The issue is that your model would have to implicitly learn that [0, 1) was one class, [1,2) is another, [2,3) is another, etc. It's impossible for the network to express a case where it's 40% confident the pixel belongs to class 0, and 60% confident the pixel belongs to class C. You would also find it impossible to apply a loss function, because your values represent classes and not values. For example, say we have 3 classes, dog, cat, and mouse. Say we also have the model predict HxW with values in [0, C], and our loss function is the absolute difference between label and prediction. Consider a single pixel, with ground truth label 2 (mouse). If our model predicts 0 for dog, the loss is 2. However, if the model predicts cat (which is equally wrong), the loss is 1. Predicting in the range [0, C] doesn't work, you must one-hot encode.

**Comment by julienguegan** [comment]
OK thank you for your anwser, my global understanding is that it depends of what loss you are using and how is it implemented ...
